### PR TITLE
Add support for variable batch size debug mode and NVTX ranges

### PR DESCRIPTION
**Pull request by ksztenderski**
## Category:

**New feature** (*non-breaking change which adds functionality*)

## Description:

Adds support for NVTX ranges in eager operators.
Adds support for iteration batch size based on outputs from `external_source`.
The support is limited to cases when `external_source` is the first operator in the pipeline.

Why is it impossible/inconvenient to support `external_source` that is defined later?

For this to work we would have to have similar execution flow to the one from standard mode (fetching data from `external_source` at the start of each run). This means we would have to first build the pipeline to save all `external_source` callbacks. The problem with this approach is that because debug mode allows to access and modify the data inside the pipeline during the build we have to operate on the actual data, we cannot have some placeholders like in standard build.

Why can we not run the pipeline once to build the graph and collect `external_source` operators?

- We cannot do this in the pipeline build because `external_source` may not be fed with data.
- In the first iteration we could run the pipeline twice, once to build the graph and collect ESs, and then a regular run. That way we would introduce some possible garbage outputs from the build phase and would have to account for all possible iteration-based states (maybe someone uses some clear python code inside our pipeline that is dependent on the number of executions) and it would probably be impossible to actually catch all of these cases.

### Affected modules and functionalities:

debug mode pipeline

### Tests
- [x] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by ksztenderski** [comment]
!build

File: dali/test/python/test_pipeline_debug.py [hunk 1]
```diff
@@ -533,3 +516,70 @@ def test_multiple_input_sets():
     pipe_standard = multiple_input_sets_pipeline()
     pipe_debug = multiple_input_sets_pipeline(debug=True)
     compare_pipelines(pipe_standard, pipe_debug, 8, 10)
+
+
+@pipeline_def(batch_size=8, num_threads=3, device_id=0, seed=47, debug=True)
+def variable_batch_size_from_external_source_pipeline(variable_batch_size):
```
Comment (klecki) [inline dali/test/python/test_pipeline_debug.py:522]: This just tests one iteration to be smaller than the max_batch_size, but not exactly changing it iter2iter, right?

File: dali/pipeline/operator/eager_operator.h [hunk 2]
```diff
@@ -123,24 +140,26 @@ template <>
 std::vector<std::shared_ptr<TensorList<CPUBackend>>> EagerOperator<CPUBackend>::Run(
     const std::vector<std::shared_ptr<TensorList<CPUBackend>>> &inputs,
     const std::unordered_map<std::string, std::shared_ptr<TensorList<CPUBackend>>> &kwargs,
-    ThreadPool *thread_pool) {
+    ThreadPool *thread_pool, int batch_size) {
+  DomainTimeRange tr("[DALI][CPU op] " + name_, DomainTimeRange::kBlue1);
```
[the remark this answers is not in the document]
Comment (ksztenderski) [reply dali/pipeline/operator/eager_operator.h:144]: done

File: dali/pipeline/pipeline_debug.h [hunk 3]
```diff
@@ -44,6 +44,7 @@ class DLL_PUBLIC PipelineDebug {
 
   DLL_PUBLIC void AddOperator(OpSpec &spec, int logical_id) {
     FillOpSpec(spec);
+    std::string op_name = "__debug__" + spec.name() + "_" + std::to_string(logical_id);
```
[the remark this answers is not in the document]
Comment (ksztenderski) [reply dali/pipeline/pipeline_debug.h:47]: Removed (an artifact of an earlier version).

File: dali/pipeline/operator/eager_operator.h [hunk 4]
```diff
@@ -55,8 +59,18 @@ template <typename Backend>
...
  public:
   DLL_PUBLIC inline EagerOperator(const OpSpec &spec)
-      : batch_size_(spec.GetArgument<int>("max_batch_size")),
+      : max_batch_size_(spec.GetArgument<int>("max_batch_size")),
+        op_spec_(spec),
+        name_(spec.name()),
+        op_(InstantiateOperator(spec)) {
+    num_outputs_ = op_spec_.GetSchema().CalculateOutputs(op_spec_) +
+                   op_spec_.GetSchema().CalculateAdditionalOutputs(op_spec_);
+  }
```
[the remark this answers is not in the document]
Comment (ksztenderski) [reply dali/pipeline/operator/eager_operator.h:68]: Of course, done. It was a quick check and forgot to change that.

File: dali/test/python/test_pipeline_debug.py (hunk 1 above)
Comment (ksztenderski) [reply dali/test/python/test_pipeline_debug.py:522]: Yes, added more iterations with changing batch size

File: dali/pipeline/operator/eager_operator.h [hunk 5]
```diff
@@ -163,35 +183,56 @@ template <>
...
+      batch_size = cur_batch_size;
+    }
+
+    DALI_ENFORCE(cur_batch_size == batch_size,
+                 make_string("Expected uniform batch size in a single operator. Expected: ",
+                             batch_size, ", got: ", cur_batch_size));
+    DALI_ENFORCE(
+        cur_batch_size <= max_batch_size_,
+        make_string("Expected batch size lower or equal to max batch size. Expected at most: ",
+                    max_batch_size_, ", got: ", batch_size));
```
[the remark this answers is not in the document]
Comment (ksztenderski) [reply dali/pipeline/operator/eager_operator.h:234]: done

File: dali/python/nvidia/dali/_debug_mode.py [hunk 6]
```diff
@@ -533,6 +549,7 @@ def run(self):
 
         self._debug_on = True
         self._cur_operator_id = -1
+        self._cur_batch_info = _BatchInfo(-1, None)  # Used for variable batch sizes.
```
[the remark this answers is not in the document]
Comment (ksztenderski) [reply dali/python/nvidia/dali/_debug_mode.py:552]: done

File: dali/python/nvidia/dali/_debug_mode.py [hunk 7]
```diff
@@ -579,19 +597,44 @@ def feed_input(self, data_node, data, **kwargs):
...
+                self._cur_batch_info = _BatchInfo(batch_size, cur_context)
+            elif self._cur_batch_info.size != batch_size:
+                if self._cur_batch_info.source_context == cur_context and output_idx > 0:
+                    raise RuntimeError(("External source must return outputs with consistent batch size. Output {} "
+                                        "has batch size = {}, previous batch size = {}")
+                                       .format(output_idx, batch_size, self._cur_batch_info.size))
+                else:
+                    raise RuntimeError(
+                        ("Batch size must be uniform across an iteration. External Source operator returned batch with "
+                         "size = {}, for this iteration used batch size = {}.\nIf you want to use batch size returned by "
```
[the remark this answers is not in the document]
Comment (ksztenderski) [reply dali/python/nvidia/dali/_debug_mode.py]: done

**Comment by ksztenderski** [comment]
!build

File: dali/python/nvidia/dali/_debug_mode.py [hunk 8]
```diff
@@ -503,6 +547,7 @@ def __init__(self, exec_func, **kwargs):
         self._next_logical_id = 0
         self._operators = {}
         self._operators_built = False
+        self._cur_iter_batch_info = _IterBatchInfo(-1, None)  # Used for variable batch sizes.
```
Comment (klecki) [inline dali/python/nvidia/dali/_debug_mode.py:550]: Nitpick:
```suggestion
        self._cur_iter_batch_info.reset()  # Used for variable batch sizes.
```

**Review by klecki — APPROVED** [review]

File: dali/python/nvidia/dali/_debug_mode.py (hunk 8 above)
Comment (klecki) [reply dali/python/nvidia/dali/_debug_mode.py:550]: Ok, it's actual constructor.

**Review by banasraf — APPROVED** [review]

**Pull request merged**